File: src/test/java/com/memverge/mvfs/LogListener.java
package com.memverge.mvfs;

import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.apache.commons.lang3.StringUtils;
import org.testng.ITestResult;
import org.testng.TestListenerAdapter;

@Slf4j
public class LogListener extends TestListenerAdapter {

  @Override
  public void onTestStart(ITestResult tr) {
    super.onTestStart(tr);
    log.info("--- {}{} --- test start.", tr.getName(), getParams(tr));
  }

  @Override
  public void onTestFailure(ITestResult tr) {
    log.error("--- {}{} --- failed, took {}s.",
        tr.getName(), getParams(tr), getSeconds(tr));
    val params = tr.getParameters();
    if (params.length > 0) {
      log.error("test parameters: {}.", params);
    }
    log.error("detail:", tr.getThrowable());
  }

  @Override
  public void onTestSkipped(ITestResult tr) {
    log.warn("--- {}{} --- skipped, took {}s.",
        tr.getName(), getParams(tr), getSeconds(tr));
  }

  @Override
  public void onTestSuccess(ITestResult tr) {
    log.info("--- {}{} --- passed, took {}s.",
        tr.getName(), getParams(tr), getSeconds(tr));
  }

  private String getParams(ITestResult tr) {
    val params = tr.getParameters();
    String ret = "";
    if (params.length > 0) {
      ret = String.format(" [%s]", StringUtils.join(params, ", "));
    }
    return ret;
  }

  private double getSeconds(ITestResult tr) {
    return (double) (tr.getEndMillis() - tr.getStartMillis()) / 1000;
  }
}

File: src/test/scala/org/apache/spark/shuffle/LogListener.scala
/*
 * Copyright (C) 2018 MemVerge Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.spark.shuffle

import org.apache.commons.lang3.StringUtils
import org.apache.spark.internal.Logging
import org.testng.{ITestResult, TestListenerAdapter}

class LogListener extends TestListenerAdapter with Logging {
  override def onTestStart(tr: ITestResult): Unit = {
    super.onTestStart(tr)
    logInfo(s"--- ${tr.getName}${getParams(tr)} --- test start.")
  }

  override def onTestFailure(tr: ITestResult): Unit = {
    logError(s"--- ${tr.getName}${getParams(tr)} --- failed, took ${getSeconds(tr)}s.")
    val params = tr.getParameters
    if (params.nonEmpty) {
      logError(s"test parameters: $params.")
    }
    logError("detail:", tr.getThrowable)
  }

  override def onTestSkipped(tr: ITestResult): Unit = {
    logWarning(s"--- ${tr.getName}${getParams(tr)} --- skipped, took ${getSeconds(tr)}s.")
  }

  override def onTestSuccess(tr: ITestResult): Unit = {
    logInfo(s"--- ${tr.getName}${getParams(tr)} --- passed, took ${getSeconds(tr)}s.")
  }

  private def getParams(tr: ITestResult): String = {
    val params = tr.getParameters
    if (params.nonEmpty) {
      s" [${StringUtils.join(params, ", ")}]"
    } else {
      ""
    }
  }

  private def getSeconds(tr: ITestResult) = (tr.getEndMillis - tr.getStartMillis).toDouble / 1000
}
